perf/collector_libpfm.go

```diff
@@ -152,68 +184,145 @@ func (c *collector) setup() error {
 	c.cpuFilesLock.Lock()
 	defer c.cpuFilesLock.Unlock()
 	cgroupFd := int(cgroup.Fd())
-	for _, group := range c.events.Core.Events {
-		customEvent, ok := c.eventToCustomEvent[group[0]]
-		var err error
-		if ok {
-			err = c.setupRawNonGrouped(customEvent, cgroupFd)
-		} else {
-			err = c.setupNonGrouped(string(group[0]), cgroupFd)
+	for i, group := range c.events.Core.Events {
+		// CPUs file descriptors of group leader needed for perf_event_open.
+		leaderFileDescriptors := make(map[int]int, len(c.onlineCPUs))
+		for _, cpu := range c.onlineCPUs {
+			leaderFileDescriptors[cpu] = groupLeaderFileDescriptor
 		}
-		if err != nil {
-			return err
+
+		for j, event := range group.events {
+			// First element is group leader.
+			isGroupLeader := j == 0
+			customEvent, ok := c.eventToCustomEvent[event]
+			if ok {
+				config := c.createConfigFromRawEvent(customEvent, isGroupLeader)
+				leaderFileDescriptors, err = c.registerEvent(eventInfo{string(customEvent.Name), config, cgroupFd, i}, leaderFileDescriptors)
+				if err != nil {
+					return err
+				}
+			} else {
+				config, err := c.createConfigFromEvent(event, isGroupLeader)
+				if err != nil {
+					return err
+				}
+				leaderFileDescriptors, err = c.registerEvent(eventInfo{string(event), config, cgroupFd, i}, leaderFileDescriptors)
+				if err != nil {
+					return err
+				}
+				// Clean memory allocated by C code.
+				C.free(unsafe.Pointer(config))
+			}
+		}
+
+		// Group is prepared so we should reset and enable counting.
+		for _, fd := range leaderFileDescriptors {
+			err = unix.IoctlSetInt(fd, unix.PERF_EVENT_IOC_RESET, 0)
+			if err != nil {
+				return err
+			}
+			err = unix.IoctlSetInt(fd, unix.PERF_EVENT_IOC_ENABLE, 0)
+			if err != nil {
+				return err
+			}
 		}
 	}
 
 	return nil
 }
 
-func (c *collector) setupRawNonGrouped(event *CustomEvent, cgroup int) error {
-	klog.V(5).Infof("Setting up non-grouped raw perf event %#v", event)
-	config := createPerfEventAttr(*event)
-	err := c.registerEvent(config, string(event.Name), cgroup)
-	if err != nil {
-		return err
+func readPerfEventAttr(name string, perfEventAttrMemory unsafe.Pointer) error {
```

Comment:
Instead of using a return argument you should simply return value from the function and allocate memory for `unix.PerfEventAttr` in this function. It took me a bit too much time to understand what's going on here.
```golang
perfEventAttrMemory := C.malloc(C.ulong(unsafe.Sizeof(unix.PerfEventAttr{})))
...
return (*unix.PerfEventAttr)(perfEventAttrMemory)
```